src/server/game/Server/WorldSocket.cpp

```diff
@@ -360,6 +360,29 @@ WorldSocket::ReadDataHandlerResult WorldSocket::ReadDataHandler()
             if (_worldSession)
                 _worldSession->ResetTimeOutTime(true);
             break;
+
+        case CMSG_TIME_SYNC_RESP:
```

Comment:
Feedback on this touch part would be nice

Reply:
It's not good actually. You have a race on `WorldSession::timeSyncClockDelta` in two threads - the network thread that this socket is bound to and any map update thread where movement packets are processed

I assume you need the timestamp of when this packet was received to be as accurate as possible (without the delay from packet queue) so my suggestion would be to somehow pass the timestamp to regular packet queue, possibly by extending WorldPacket class. Marking some opcodes as "high priority" would get the timestamp with `getMSTime()` while regular ones would use lower precision `GameTime::GetGameTimeMS()`

Reply:
> It's not good actually. You have a race on WorldSession::timeSyncClockDelta in two threads - the network thread that this socket is bound to and any map update thread where movement packets are processed

Indeed. Wouldn't using `std::atomic<int64> timeSyncClockDelta` instead of `int64 timeSyncClockDelta` fix the issue though? This is the way `CMSG_PING` and `CMSG_KEEP_ALIVE` are handled.

Though, adding a received timestamp in WorldPacket (or a subclass), as suggested by you and @akrom23, would work too and would be cleaner I suppose.

Reply:
Done. The time when `CMSG_TIME_SYNC_RESP` packets are received are now recorded (before they are put into the packet queue). And then, they are handled like any other `PROCESS_THREADSAFE` packet.